bindings/python/soletta/mainloop.pyx

```diff
@@ -0,0 +1,90 @@
+import asyncio
+from libc.stdint cimport uint32_t
+
+cdef extern from "sol-mainloop.h":
+	struct sol_mainloop_implementation:
+		void run()
+		void init()
+		void quit()
+		void * idle_add(bint (*)(void*), void*)
+		bint idle_del(void*)
+		void* timeout_add(uint32_t timeout_ms, bint (*cb)(void *), const void *)
+		bint timeout_del(void*)
+		void shutdown()
+		pass
+
+	bint sol_mainloop_set_implementation(const sol_mainloop_implementation *)
+
+
+cdef void wrap_sol_init():
+	pass
+
+cdef void wrap_sol_shutdown():
+	pass
+
+cdef void wrap_sol_quit():
+	asyncio.get_event_loop().stop()
+
+cdef void wrap_sol_run():
+	asyncio.get_event_loop().run_forever()
+
+cdef void *wrap_sol_timeout_add(uint32_t timeout_ms, bint (*cb)(void *), const void *data):
+	
+	@asyncio.coroutine
+	def task():
+		while cb(data):
+			yield from asyncio.sleep(timeout_ms / 1000)
+
+	t = asyncio.get_event_loop().create_task(task())
+	return <void*> t
+
+cdef bint wrap_sol_timeout_del(void *handle):
+	(<object?>handle).cancel()
+	return True
+
+cdef void *wrap_idle_add(bint (*cb)(void*), const void *data ):
+
+	@asyncio.coroutine
+	def task():
+		while True:
+			cb(data)
+			"""FIXME: this is probably not what we want"""
+			yield from asyncio.sleep(0.1)
```

Comment:
AFAIK, asyncio doesn't have the sort of idle handler concept that exists in glib and others.  The typical pattern that is somewhat similar is to write a coroutine that yields to the event loop for some period of time.  The lower the priority, the longer the coroutine can suspend while the event loop continues with other tasks.

This allows for somewhat more deterministic behavior then what you'd get with glib.